torch/_subclasses/fake_tensor.py

```diff
@@ -730,6 +728,27 @@ def multi_device_op_out(fake_mode, func, *args, **kwargs):
     return new_kwargs["input"]
 
 
+@register_op_impl(aten.index_put.default)
+@register_op_impl(aten.index_put_.default)
+def index_put_impl(fake_mode, func, *args, **kwargs):
+    _, new_kwargs = normalize_function(
+        func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
+    )
+
+    values = new_kwargs["values"]
+    self_device = new_kwargs["input"].fake_device
+    torch._check(
+        self_device == values.fake_device or (values.ndim == 0 and values.numel() == 1),
+        lambda: f"Mismatching {func} device between self ({self_device}) and values ({values.device})",
```

Comment:
just curious - why does this get special treatment in fake tensor instead of `index_put_` having a meta impl in `_meta_registrations.py`?

Reply:
It's device error checking. We could move more of that into meta_registration, similar to how we check device_hint. I'm not sure - which do you think would be better ?

Reply:
My mental model was always "fake/meta impls live outside this file, and the only stuff that lives in this file is:

(1) core FakeTensorMode machinery

(2) fake/meta impls that require special treatment (e.g. factory functions, which require using `in_kernel_invocation_manager`)

So whenever I see a meta impl in this file I think "it must be doing something special". So I guess I would lightly prefer for "normal" meta impls to live outside this file, but I don't have a strong preference

Reply:
In this case, it's something special because it has non default device behavior.